File: android/app/src/main/java/com/aurea/aurea/editor/CommandSearch.kt
package com.aurea.aurea.editor

import android.content.Context
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.aurea.aurea.editor.panels.*
import com.aurea.aurea.presets.PresetEntry
import com.aurea.aurea.presets.PresetKind
import com.aurea.aurea.state.EditorStore
import com.aurea.aurea.ui.theme.AureaColors
import org.json.JSONArray

internal data class EditorCommand(val id: String, val title: String, val detail: String, val keywords: String, val requires: String)
private data class CommandHit(val id: String, val title: String, val detail: String, val search: String,
                              val category: String, val requires: String, val effect: Int? = null, val preset: PresetEntry? = null)

// Stable IDs and capabilities come from the same resource on both platforms.
internal fun readEditorCommands(context: Context): List<EditorCommand> {
    val array = JSONArray(context.assets.open("editor_commands.json").bufferedReader().use { it.readText() })
    return (0 until array.length()).map { i -> array.getJSONObject(i).let {
        EditorCommand(it.getString("id"), it.getString("title"), it.getString("detail"), it.getString("keywords"), it.getString("requires"))
    } }
}

private fun commandUnavailable(requirement: String, store: EditorStore): String? {
    val selected = store.layers.filter { it.id in store.selection }
    if (requirement == "any") return null
    if (requirement == "layers") return if (store.layers.isNotEmpty()) null else "Adicione uma camada"
    if (requirement == "undo") return if (store.project.canUndo) null else "Nenhuma edição para desfazer"
    if (requirement == "redo") return if (store.project.canRedo) null else "Nenhuma edição para refazer"
    if (selected.isEmpty()) return "Selecione uma camada"
    if (selected.any { it.locked }) return "Desbloqueie a seleção para editar"
    if (requirement == "selection") return null
    if (requirement == "inside") return if (selected.any { store.playhead > it.startFrame && store.playhead < it.endFrame }) null else "Posicione o cabeçote dentro do clipe"
    val layer = selected.singleOrNull() ?: return "Selecione apenas uma camada"
    return when (requirement) {
        "single" -> null
        "inside_single" -> if (store.playhead > layer.startFrame && store.playhead < layer.endFrame) null else "Posicione o cabeçote dentro do clipe"
        "video", "video_inside" -> if (layer.kind != 1) "Selecione um vídeo" else if (requirement == "video_inside" && store.playhead !in layer.startFrame until layer.endFrame) "Posicione o cabeçote dentro do vídeo" else null
        "media" -> if (layer.kind in listOf(1, 3)) null else "Selecione vídeo ou áudio"
        "text" -> if (layer.kind == 4) null else "Selecione um texto"
        "visual" -> if (layer.kind != 3) null else "Selecione uma camada visual"
        "text3d" -> if (store.text3d != null) null else "Selecione um texto 3D"
        "model3d" -> if (layer.kind == 10) null else "Selecione um modelo 3D"
        "particles" -> if (layer.kind == 11) null else "Selecione partículas"
        "vector" -> if (store.isVectorLayer) null else "Selecione um vetor"
        "precomp" -> if (layer.kind == 12) null else "Selecione uma pré-composição"
        else -> "Ação indisponível neste contexto"
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun CommandSearchSheet(store: EditorStore, ui: EditorUi, onDismiss: () -> Unit) {
    val context = LocalContext.current
    val commands = remember { readEditorCommands(context) }
    val preferences = remember { context.getSharedPreferences("aurea.commands", Context.MODE_PRIVATE) }
    var favorites by remember { mutableStateOf(preferences.getStringSet("favorites", emptySet()).orEmpty().toSet()) }
    var query by rememberSaveable { mutableStateOf("") }
    var category by rememberSaveable { mutableStateOf("Tudo") }
    val presets = store.presets.all()
    // Reuse existing catalogs and preference stores; no duplicate effect/preset system.
    val hits = commands.map { CommandHit(it.id, it.title, it.detail, normalizeSearch("${it.title} ${it.keywords} ${it.detail}"), "Ações", it.requires) } +
        store.catalog.map { CommandHit("effect:${it.typeId}", effectDisplayName(it.typeId, it.name), "Adicionar efeito · ${it.category}",
            effectSearchText(it.typeId, it.name, it.category), "Efeitos", if (it.typeId == effectTypeId("aurea.text3d.layout")) "text3d" else "selection", effect = it.typeId) } +
        presets.map { CommandHit("preset:${it.key}", it.name, "Abrir preset · ${it.kind.dir}", normalizeSearch("${it.name} ${it.kind.dir} preset"),
            "Presets", if (it.kind == PresetKind.Text) "text" else "single", preset = it) }
    fun isFavorite(hit: CommandHit) = hit.effect?.let { store.effectPrefs.isFavorite(it) }
        ?: hit.preset?.let { it.key in store.presets.favorites } ?: (hit.id in favorites)
    val terms = normalizeSearch(query).split(Regex("\\s+")).filter { it.isNotBlank() }
    val filtered = hits.filter { hit ->
        (category == "Tudo" || category == hit.category || category == "Favoritos" && isFavorite(hit)) &&
            terms.all { it in hit.search } &&
            (terms.isNotEmpty() || category != "Tudo" || isFavorite(hit) || hit.category == "Ações" && commandUnavailable(hit.requires, store) == null)
    }.sortedByDescending { isFavorite(it) }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true), containerColor = AureaColors.EditorPanel) {
        Column(Modifier.fillMaxWidth().fillMaxHeight(0.88f).padding(horizontal = 16.dp).imePadding()) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text("Ferramentas", color = AureaColors.Text, fontSize = 20.sp, fontWeight = FontWeight.SemiBold, modifier = Modifier.weight(1f))
                TextButton(onClick = onDismiss) { Text("Fechar") }
            }
            OutlinedTextField(query, { query = it }, modifier = Modifier.fillMaxWidth().semantics { contentDescription = "Buscar ferramentas" },
                singleLine = true, placeholder = { Text("Buscar ação, efeito ou preset") })
            androidx.compose.foundation.lazy.LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                items(listOf("Tudo", "Favoritos", "Ações", "Efeitos", "Presets")) { name -> FilterChip(selected = category == name, onClick = { category = name }, label = { Text(name) }) }
            }
            if (filtered.isEmpty()) Text(if (category == "Favoritos") "Toque na estrela de uma ferramenta para guardar aqui." else "Nenhuma ferramenta encontrada.", color = AureaColors.Muted, modifier = Modifier.padding(vertical = 20.dp))
            LazyColumn(Modifier.weight(1f)) {
                items(filtered, key = { it.id }) { hit ->
                    val reason = commandUnavailable(hit.requires, store)
                    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                        Column(Modifier.weight(1f).heightIn(min = 64.dp).clickable(enabled = reason == null) {
                            // Resolve against the current selection, never a captured layer ID.
                            if (commandUnavailable(hit.requires, store) == null) {
                                onDismiss()
                                when {
                                    hit.effect != null -> { store.addEffect(hit.effect); store.effectPrefs.addRecent(hit.effect); openPanel(store, ui, EditorPanel.Effects) }
                                    hit.preset != null -> { store.presetsOpenKind = hit.preset.kind; store.presetsOpenSearch = hit.preset.name; openPanel(store, ui, EditorPanel.Presets) }
                                    else -> executeEditorCommand(hit.id, store, ui)
                                }
                            }
                        }.padding(vertical = 10.dp, horizontal = 4.dp)) {
                            Text(hit.title, color = if (reason == null) AureaColors.Text else AureaColors.Muted, fontSize = 15.sp)
                            Text(reason ?: hit.detail, color = AureaColors.Muted, fontSize = 12.sp)
                        }
                        TextButton(onClick = {
                            when {
                                hit.effect != null -> store.effectPrefs.toggleFavorite(hit.effect)
                                hit.preset != null -> store.presets.toggleFavorite(hit.preset)
                                else -> { favorites = if (hit.id in favorites) favorites - hit.id else favorites + hit.id; preferences.edit().putStringSet("favorites", favorites).apply() }
                            }
                        }, modifier = Modifier.size(48.dp).semantics { contentDescription = "${if (isFavorite(hit)) "Remover dos" else "Adicionar aos"} favoritos: ${hit.title}" }) {
                            Text(if (isFavorite(hit)) "★" else "☆", color = AureaColors.Accent, fontSize = 24.sp)
                        }
                    }
                }
            }
        }
    }
}

private fun executeEditorCommand(id: String, store: EditorStore, ui: EditorUi) {
    val layer = store.primary
    fun panel(value: EditorPanel) = openPanel(store, ui, value)
    when (id) {
        "split" -> store.splitAtPlayhead()
        "duplicate" -> store.duplicateLayers()
        "ripple_delete" -> store.deleteLayers(ripple = true)
        "trim_start" -> layer?.let { store.trimStart(it, store.playhead) }
        "trim_end" -> layer?.let { store.trimEnd(it, store.playhead) }
        "speed" -> panel(EditorPanel.Speed)
        "clip_edit" -> panel(EditorPanel.ClipEdit)
        "freeze" -> store.freezeFrame()
        "extract_audio" -> store.extractAudio()
        "audio" -> panel(EditorPanel.Audio)
        "beats" -> store.detectBeats()
        "transform" -> panel(EditorPanel.Transform)
        "text" -> panel(EditorPanel.Text)
        "mask" -> panel(EditorPanel.Mask)
        "effects" -> panel(EditorPanel.Effects)
        "appearance" -> panel(EditorPanel.Appearance)
        "tracking" -> panel(EditorPanel.Tracking)
        "environment" -> panel(EditorPanel.Element3D)
        "particles" -> panel(EditorPanel.Particles)
        "vector" -> panel(EditorPanel.Vector)
        "presets" -> panel(EditorPanel.Presets)
        "precompose" -> store.precompose()
        "enter_precomp" -> layer?.let { store.openPrecomp(it) }
        "marker" -> store.toggleMarker()
        "magnetic" -> store.toggleEditMode()
        "remove_gaps" -> store.removeGaps()
        "previous_frame" -> store.step(-1)
        "next_frame" -> store.step(1)
        "add_text" -> store.addText()
        "add_null" -> store.addNull(false)
        "add_null3d" -> store.addNull(true)
        "add_camera" -> store.addCamera()
        "captions" -> panel(EditorPanel.Captions)
        "project_settings" -> openSheet(store, ui, ShellSheet.ProjectSettings)
        "search_layers" -> openSheet(store, ui, ShellSheet.SearchLayers)
        "undo" -> store.undo()
        "redo" -> store.redo()
    }
}

File: engine/platform/ios/app/CommandSearch.swift
import SwiftUI

struct EditorCommand: Decodable, Identifiable {
    let id: String
    let title: String
    let detail: String
    let keywords: String
    let requires: String
    static let catalog: [EditorCommand] = {
        guard let url = Bundle.main.url(forResource: "editor_commands", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let values = try? JSONDecoder().decode([EditorCommand].self, from: data) else { return [] }
        return values
    }()
}

private struct CommandHit: Identifiable {
    let id: String
    let title: String
    let detail: String
    let search: String
    let category: String
    let requires: String
    var effect: UInt32?
    var preset: PanelPresetEntry?
}

@MainActor
struct CommandSearchView: View {
    @EnvironmentObject private var model: AureaModel
    @EnvironmentObject private var shell: ShellPresentation
    @Environment(\.dismiss) private var dismiss
    @StateObject private var effectPrefs = FxEffectPrefs()
    @State private var favorites = Set(UserDefaults.standard.stringArray(forKey: "aurea.commands.favorites") ?? [])
    @State private var presetFavorites = Set(UserDefaults.standard.stringArray(forKey: "presetFavorites") ?? [])
    @State private var query = ""
    @State private var category = "Tudo"
    @State private var hits: [CommandHit] = []

    private func unavailable(_ requirement: String) -> String? {
        let selected = model.layers.filter { model.selection.contains($0.id) }
        if requirement == "any" { return nil }
        if requirement == "layers" { return model.layers.isEmpty ? "Adicione uma camada" : nil }
        if requirement == "undo" { return model.status.canUndo == 0 ? "Nenhuma edição para desfazer" : nil }
        if requirement == "redo" { return model.status.canRedo == 0 ? "Nenhuma edição para refazer" : nil }
        if selected.isEmpty { return "Selecione uma camada" }
        if selected.contains(where: { $0.locked }) { return "Desbloqueie a seleção para editar" }
        if requirement == "selection" { return nil }
        if requirement == "inside" {
            return selected.contains { model.status.playhead > Int64($0.startFrame) && model.status.playhead < Int64($0.endFrame) } ? nil : "Posicione o cabeçote dentro do clipe"
        }
        guard selected.count == 1, let layer = selected.first else { return "Selecione apenas uma camada" }
        switch requirement {
        case "single": return nil
        case "inside_single": return model.status.playhead > Int64(layer.startFrame) && model.status.playhead < Int64(layer.endFrame) ? nil : "Posicione o cabeçote dentro do clipe"
        case "video", "video_inside":
            if layer.kind != 1 { return "Selecione um vídeo" }
            return requirement == "video_inside" && (model.status.playhead < Int64(layer.startFrame) || model.status.playhead >= Int64(layer.endFrame)) ? "Posicione o cabeçote dentro do vídeo" : nil
        case "media": return [1, 3].contains(layer.kind) ? nil : "Selecione vídeo ou áudio"
        case "text": return layer.kind == 4 ? nil : "Selecione um texto"
        case "visual": return layer.kind != 3 ? nil : "Selecione uma camada visual"
        case "model3d": return layer.kind == 10 ? nil : "Selecione um modelo 3D"
        case "particles": return layer.kind == 11 ? nil : "Selecione partículas"
        case "vector": return model.isVectorLayer ? nil : "Selecione um vetor"
        case "precomp": return layer.kind == 12 ? nil : "Selecione uma pré-composição"
        default: return "Ação indisponível neste contexto"
        }
    }

    private func favorite(_ hit: CommandHit) -> Bool {
        if let effect = hit.effect { return effectPrefs.isFavorite(effect) }
        if let preset = hit.preset { return presetFavorites.contains(preset.id) }
        return favorites.contains(hit.id)
    }
    private var results: [CommandHit] {
        let terms = fxNormalizeSearch(query).split(whereSeparator: { $0.isWhitespace }).map(String.init)
        return hits.filter { hit in
            (category == "Tudo" || category == hit.category || category == "Favoritos" && favorite(hit)) &&
            terms.allSatisfy { hit.search.contains($0) } &&
            (!terms.isEmpty || category != "Tudo" || favorite(hit) || hit.category == "Ações" && unavailable(hit.requires) == nil)
        }.enumerated().sorted { a, b in
            let af = favorite(a.element), bf = favorite(b.element)
            return af == bf ? a.offset < b.offset : af
        }.map(\.element)
    }

    var body: some View {
        VStack(spacing: 10) {
            HStack {
                Text("Ferramentas").font(.aurea(size: 20, weight: .semibold))
                Spacer()
                Button("Fechar") { dismiss() }.frame(minWidth: 44, minHeight: 44)
            }
            TextField("Buscar ação, efeito ou preset", text: $query)
                .textInputAutocapitalization(.never).autocorrectionDisabled()
                .padding(12).background(AureaColors.chip, in: RoundedRectangle(cornerRadius: 10))
                .accessibilityIdentifier("commandSearchField")
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(["Tudo", "Favoritos", "Ações", "Efeitos", "Presets"], id: \.self) { name in
                        Button { category = name } label: {
                            Text(name).font(.aurea(size: 13)).padding(.horizontal, 12).frame(minHeight: 44)
                                .background(category == name ? AureaColors.accentDim : AureaColors.chip, in: Capsule())
                        }.buttonStyle(.plain)
                    }
                }
            }
            if results.isEmpty {
                Text(category == "Favoritos" ? "Toque na estrela de uma ferramenta para guardar aqui." : "Nenhuma ferramenta encontrada.")
                    .foregroundStyle(AureaColors.muted).padding(.vertical, 20)
            }
            ScrollView {
                LazyVStack(spacing: 0) {
                    ForEach(results) { hit in commandRow(hit) }
                }
            }
        }
        .padding(16).foregroundStyle(AureaColors.text).background(AureaColors.editorPanel)
        .onAppear(perform: load)
    }

    private func commandRow(_ hit: CommandHit) -> some View {
        let reason = unavailable(hit.requires)
        return HStack(spacing: 0) {
            Button {
                guard unavailable(hit.requires) == nil else { return }
                dismiss()
                execute(hit)
            } label: {
                VStack(alignment: .leading, spacing: 4) {
                    Text(hit.title).font(.aurea(size: 15)).foregroundStyle(reason == nil ? AureaColors.text : AureaColors.muted)
                    Text(reason ?? hit.detail).font(.aurea(size: 12)).foregroundStyle(AureaColors.muted)
                }.padding(.vertical, 10).frame(maxWidth: .infinity, minHeight: 64, alignment: .leading).contentShape(Rectangle())
            }.buttonStyle(.plain).disabled(reason != nil).accessibilityIdentifier("command:\(hit.id)")
            Button { toggleFavorite(hit) } label: {
                Image(systemName: favorite(hit) ? "star.fill" : "star").foregroundStyle(AureaColors.accent).frame(width: 48, height: 48)
            }.buttonStyle(.plain).accessibilityLabel("\(favorite(hit) ? "Remover dos" : "Adicionar aos") favoritos: \(hit.title)")
        }
    }

    private func load() {
        hits = EditorCommand.catalog.map {
            CommandHit(id: $0.id, title: $0.title, detail: $0.detail, search: fxNormalizeSearch("\($0.title) \($0.detail) \($0.keywords)"), category: "Ações", requires: $0.requires)
        }
        hits += model.effectCatalog.map {
            CommandHit(id: "effect:\($0.typeId)", title: fxEffectDisplayName($0.typeId, $0.name), detail: "Adicionar efeito · \($0.category)", search: fxEffectSearchText($0.typeId, $0.name, $0.category), category: "Efeitos", requires: "selection", effect: $0.typeId)
        }
        hits += PanelPresetEntry.loadAll().map {
            CommandHit(id: "preset:\($0.id)", title: $0.name, detail: "Abrir preset · \($0.kind.rawValue)", search: fxNormalizeSearch("\($0.name) \($0.kind.rawValue) preset"), category: "Presets", requires: $0.kind == .text ? "text" : "single", preset: $0)
        }
    }

    private func toggleFavorite(_ hit: CommandHit) {
        if let effect = hit.effect { effectPrefs.toggleFavorite(effect); return }
        if let preset = hit.preset {
            if presetFavorites.contains(preset.id) { presetFavorites.remove(preset.id) } else { presetFavorites.insert(preset.id) }
            UserDefaults.standard.set(Array(presetFavorites), forKey: "presetFavorites")
        } else {
            if favorites.contains(hit.id) { favorites.remove(hit.id) } else { favorites.insert(hit.id) }
            UserDefaults.standard.set(Array(favorites), forKey: "aurea.commands.favorites")
        }
    }

    private func execute(_ hit: CommandHit) {
        if let effect = hit.effect {
            let targets = model.layers.filter { model.selection.contains($0.id) && !$0.locked }
            model.mutate { engine in
                engine.beginUndoGroup()
                for layer in targets { engine.addEffect(effect, toLayer: layer.id, at: UInt32.max) }
                engine.endUndoGroup()
            }
            effectPrefs.addRecent(effect); model.refreshModel(force: true); model.openPanel(.effects)
            return
        }
        if let preset = hit.preset {
            model.presetsOpenKind = preset.kind.rawValue; model.presetsOpenSearch = preset.name; model.openPanel(.presets)
            return
        }
        let layer = model.primarySelection
        switch hit.id {
        case "split": model.splitAtPlayhead(Array(model.selection))
        case "duplicate": model.engine.duplicateLayers(model.selection.map { NSNumber(value: $0) }); model.refreshModel(force: true)
        case "ripple_delete": model.deleteSelectedLayers(ripple: true)
        case "trim_start": if let layer { model.trimStart(layer, at: model.status.playhead) }
        case "trim_end": if let layer { model.trimEnd(layer, at: model.status.playhead) }
        case "speed": model.openPanel(.speed)
        case "freeze":
            if let layer {
                let created = model.engine.freezeFrame(forLayer: layer, frame: Int32(clamping: model.status.playhead), hold: Int32(max(1, model.compositionFps * 3)))
                model.refreshModel(force: true); if created >= 0 { model.select(layerId: created) }
            }
        case "extract_audio": if let layer { _ = model.engine.extractAudio(fromLayer: layer); model.refreshModel(force: true) }
        case "audio": model.openPanel(.audio)
        case "beats": model.detectBeats()
        case "transform": model.openPanel(.transform)
        case "text": model.openPanel(.text)
        case "mask": model.openPanel(.mask)
        case "effects": model.openPanel(.effects)
        case "appearance": model.openPanel(.appearance)
        case "tracking": model.openPanel(.tracking)
        case "environment": model.openPanel(.layer3D)
        case "particles": model.openPanel(.particles)
        case "vector": model.openPanel(.vector)
        case "presets": model.openPanel(.presets)
        case "precompose": model.groupSelection()
        case "enter_precomp": if let layer { model.openGroup(layer) }
        case "marker": model.toggleMarker()
        case "magnetic": model.toggleEditMode()
        case "remove_gaps": model.removeGaps()
        case "previous_frame": model.step(-1)
        case "next_frame": model.step(1)
        case "add_text": model.addText()
        case "add_null": model.addNull(threeD: false)
        case "add_null3d": model.addNull(threeD: true)
        case "add_camera": model.addCamera()
        case "captions": model.openPanel(.captions)
        case "project_settings": model.showProjectSettings = true
        case "search_layers": shell.sheet = .searchLayers
        case "undo": model.undo()
        case "redo": model.redo()
        default: break
        }
    }
}
